file_util.go

```diff
@@ -0,0 +1,142 @@
+package gosnowflake
+
+//lint:file-ignore U1000 TODO SNOW-29352
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	usr "os/user"
+	"path/filepath"
+	"strings"
+)
+
+type snowflakeFileUtil struct {
+}
+
+func (util *snowflakeFileUtil) compressFileWithGzipFromStream(srcStream **bytes.Buffer) (*bytes.Buffer, int) {
+	buf := getByteBufferContent(srcStream)
+	var c bytes.Buffer
+	w := gzip.NewWriter(&c)
+	w.Write(*buf) // write buf to gzip writer
+	w.Close()
+	return &c, c.Len()
+}
+
+func (util *snowflakeFileUtil) getDigestAndSize(src **bytes.Buffer) (string, int64) {
+	m := sha256.New()
+	buf := getByteBufferContent(src)
+	m.Write(*buf)
+	return base64.StdEncoding.EncodeToString(m.Sum(nil)), int64((*src).Len())
+}
+
+func (util *snowflakeFileUtil) getDigestAndSizeForStream(stream **bytes.Buffer) (string, int64) {
+	return util.getDigestAndSize(stream)
+}
+
+func (util *snowflakeFileUtil) getDigestAndSizeForFile(fileName string) (string, int64) {
+	src, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		panic(err)
+	}
+	buf := bytes.NewBuffer(src)
+	return util.getDigestAndSize(&buf)
+}
+
+// file metadata for PUT/GET
+type fileMetadata struct {
+	name               string
+	srcFileName        string
+	srcFileSize        int
+	stageLocationType  cloudType
+	stageInfo          execResponseStageInfo
+	srcCompressionType *compressionType
+	dstCompressionType *compressionType
+	requireCompress    bool
+	dstFileName        string
+	parallel           int64
+	sha256Digest       string
+	uploadSize         int64
+	encryptionMaterial *snowflakeFileEncryption
+	dstFileSize        int64
+	overwrite          bool
+	sfa                *snowflakeFileTransferAgent
+	client             cloudClient
+	realSrcFileName    string
+	tmpDir             string
+	resStatus          resultStatus
+
+	/* streaming PUT */
+	srcStream     *bytes.Buffer
+	realSrcStream *bytes.Buffer
+
+	putCallback             string
+	putAzureCallback        string
+	putCallbackOutputStream string
+	getCallback             string
+	getAzureCallback        string
+	getCallbackOutputStream string
+	presignedURL            string
+	errorDetails            error
+	lastError               error
+	noSleepingTime          int64
+	lastMaxConcurrency      int
+	localLocation           string
+	localDigest             string
+}
+
+type fileTransferResultType struct {
+	name               string
+	srcFileName        string
+	dstFileName        string
+	srcFileSize        int
+	dstFileSize        int64
+	srcCompressionType *compressionType
+	dstCompressionType *compressionType
+	resStatus          resultStatus
+	errorDetails       error
+}
+
+type fileHeader struct {
+	digest             string
+	contentLength      int64
+	encryptionMetadata encryptMetadata
+}
+
+func getByteBufferContent(src **bytes.Buffer) *[]byte {
+	var b bytes.Buffer
+	tee := io.TeeReader(*src, &b) // read src to buf
+	buf, err := ioutil.ReadAll(tee)
+	if err != nil {
+		return nil
```

Comment:
Shouldn't we be logging this?

Reply:
Not really, since this merely returns the byte content within the bytes.Buffer. Golang's bytes.Buffer class does not have a "seek" method so I thought the best alternative was to read and return the contents when dealing with streams.

Reply:
No I meant that if there's an error during the read, shouldn't we be logging that?